geomesa-spark/geomesa-spark-sql/src/main/scala/org/locationtech/geomesa/spark/sedona/package.scala

```diff
@@ -0,0 +1,49 @@
+/***********************************************************************
+ * Copyright (c) 2013-2022 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.spark
+
+import org.apache.sedona.sql.UDF.Catalog
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.sql.sedona_sql.strategy.join.JoinQueryDetector
+
+package object sedona {
+
+  // User can specify a common prefix of UDFs/UDAFs introduced by Apache Sedona. For example, when prefix is specified
+  // as "Sedona_", ST_Contains function from Apache Sedona will be named as "Sedona_ST_Contains". When prefix is explicitly
+  // set to empty, Apache Sedona functions will replace corresponding Spark JTS functions.
+  def sedonaUdfPrefix(sqlContext: SQLContext): String = sqlContext.getConf("spark.geomesa.sedona.udf.prefix", "sedona_")
+
+  /**
+   * Register Geometry UDTs, UDFs, UDAFs and optimization rules for Apache Sedona.
+   * @param sqlContext Spark [[SQLContext]] object
+   */
+  def initSedona(sqlContext: SQLContext): Unit = {
+    val prefix = sedonaUdfPrefix(sqlContext)
+    registerOptimizations(sqlContext)
+    registerUdfs(sqlContext, prefix)
+  }
+
+  private def registerOptimizations(sqlContext: SQLContext): Unit = {
+    Seq(SedonaGeometryLiteralRules).foreach { r =>
+      if (!sqlContext.experimental.extraOptimizations.contains(r))
+        sqlContext.experimental.extraOptimizations ++= Seq(r)
+    }
+    Seq(new JoinQueryDetector(sqlContext.sparkSession)).foreach { s =>
+      if(!sqlContext.experimental.extraStrategies.contains(s))
+        sqlContext.experimental.extraStrategies ++= Seq(s)
+    }
+  }
+
+  private def registerUdfs(sqlContext: SQLContext, prefix: String): Unit = {
+    val sparkSession = sqlContext.sparkSession
+    Catalog.expressions.foreach(f => sparkSession.sessionState.functionRegistry.createOrReplaceTempFunction(
```

Comment:
What about upgrading this expression to the new API of 3.2 (here for sedona registration and in the geomesa UDF registration):

https://github.com/apache/incubator-sedona/blob/5055078d18c687f92fcc67b7377203c501823b8a/sql/src/main/scala/org/apache/sedona/sql/UDF/UdfRegistrator.scala#L38

changing:

```
sparkSession.sessionState.functionRegistry.createOrReplaceTempFunction(
```

to:

```
sparkSession.sessionState.functionRegistry.registerFunction
```

This should already ensure compatibility with 3.2 Spark.

Reply:
Can we make that change and remain compatible with Spark 3.1 and before?  (If not, it may be possible to use reflection.)

Historically, I have been an advocate of providing backwards compatibility and limiting dependency changes.

Reply:
+1 to keeping back-compatibility with spark 3.1 and earlier. supporting spark 3.2 might be a larger change that we don't want to roll into this pr.

Reply:
Would be 3.1 compatible: https://github.com/apache/spark/blob/v3.1.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L47 as the method signature seems to be there as well.